src/Servers/Kestrel/Core/src/Internal/Infrastructure/Heartbeat.cs

```diff
@@ -27,58 +27,54 @@ public Heartbeat(IHeartbeatHandler[] callbacks, ISystemClock systemClock, IDebug
             _debugger = debugger;
             _trace = trace;
             _interval = Interval;
+            _timerThread = new Thread(state => ((Heartbeat)state).TimerLoop())
+            {
+                Name = "Kestrel Timer",
+                IsBackground = true
+            };
         }
 
         public void Start()
         {
             OnHeartbeat();
-            _timer = new Timer(OnHeartbeat, state: this, dueTime: _interval, period: _interval);
+            _timerThread.Start(this);
         }
 
-        private static void OnHeartbeat(object state)
-        {
-            ((Heartbeat)state).OnHeartbeat();
-        }
-
-        // Called by the Timer (background) thread
         internal void OnHeartbeat()
         {
             var now = _systemClock.UtcNow;
 
-            if (Interlocked.Exchange(ref _executingOnHeartbeat, 1) == 0)
-            {
-                Volatile.Write(ref _lastHeartbeatTicks, now.Ticks);
+            Volatile.Write(ref _lastHeartbeatTicks, now.Ticks);
 
-                try
-                {
-                    foreach (var callback in _callbacks)
-                    {
-                        callback.OnHeartbeat(now);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    _trace.LogError(0, ex, $"{nameof(Heartbeat)}.{nameof(OnHeartbeat)}");
-                }
-                finally
+            try
+            {
+                foreach (var callback in _callbacks)
                 {
-                    Interlocked.Exchange(ref _executingOnHeartbeat, 0);
+                    callback.OnHeartbeat(now);
                 }
             }
-            else
+            catch (Exception ex)
             {
-                if (!_debugger.IsAttached)
-                {
-                    var lastHeartbeatTicks = Volatile.Read(ref _lastHeartbeatTicks);
-                    
-                    _trace.HeartbeatSlow(TimeSpan.FromTicks(now.Ticks - lastHeartbeatTicks), _interval, now);
-                }
+                _trace.LogError(0, ex, $"{nameof(Heartbeat)}.{nameof(OnHeartbeat)}");
+            }
+
+            Thread.Sleep(_interval);
```

Comment:
Is this needed? In `TimerLoop`, which calls `OnHeartbeat` there's the sleep.

Reply:
oops 😄